packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,218 @@
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import type { ApiPromise } from '@polkadot/api'
+import type {
+  AssetDidUri,
+  CTypeHash,
+  IDelegationNode,
+  IPublicCredentialInput,
+  IPublicCredential,
+} from '@kiltprotocol/types'
+import type { GenericCall, Option, Result, u64, Vec } from '@polkadot/types'
+import type { Call, Extrinsic, Hash } from '@polkadot/types/interfaces'
+import type { ITuple } from '@polkadot/types/types'
+import type {
+  PublicCredentialError,
+  PublicCredentialsCredentialsCredential,
+  PublicCredentialsCredentialsCredentialEntry,
+} from '@kiltprotocol/augment-api'
+
+import { encode as cborEncode, decode as cborDecode } from 'cbor'
+
+import { u8aToHex } from '@polkadot/util'
+import { HexString } from '@polkadot/util/types'
+import { ConfigService } from '@kiltprotocol/config'
+import { fromChain as didFromChain } from '@kiltprotocol/did'
+import { validateUri } from '@kiltprotocol/asset-did'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { computeId } from './PublicCredential.js'
+
+export interface EncodedPublicCredential {
+  ctypeHash: CTypeHash
+  subject: AssetDidUri
+  claims: HexString
+  authorization: IDelegationNode['id'] | null
+}
+
+/**
+ * Format a [[IPublicCredentialInput]] to be used as a parameter for the blockchain API function.
+
+ * @param publicCredential The public credential to format.
+ * @returns The blockchain-formatted public credential.
+ */
+export function toChain(
+  publicCredential: IPublicCredentialInput
+): EncodedPublicCredential {
+  const { cTypeHash, claims, subject, delegationId } = publicCredential
+
+  const cborSerializedClaims = cborEncode(claims)
+
+  return {
+    ctypeHash: cTypeHash,
+    subject,
+    // Using Uint8Array directly fails to encode and decode, I guess because the api object assumes the byte array is SCALE-encoded.
+    claims: u8aToHex(new Uint8Array(cborSerializedClaims)),
+    authorization: delegationId,
+  }
+}
+
+// Flatten any nested batch calls into a single list of calls.
+function flattenCalls(api: ApiPromise, call: Call): Call[] {
+  if (
+    api.tx.utility.batch.is(call) ||
+    api.tx.utility.batchAll.is(call) ||
+    api.tx.utility.forceBatch.is(call)
+  ) {
+    // Inductive case
+    return call.args[0].flatMap((c) => flattenCalls(api, c))
+  }
+  // Base case
+  return [call]
+}
+
+// Transform a blockchain-formatted public credential into the original [[IPublicCredentialInput]].
+// It throws if what was written on the chain was garbage.
+function credentialInputFromChain(
+  credential: PublicCredentialsCredentialsCredential
+): IPublicCredentialInput {
+  validateUri(credential.subject.toUtf8())
+  return {
+    claims: cborDecode(credential.claims),
+    cTypeHash: credential.ctypeHash.toHex(),
+    delegationId: credential.authorization.unwrapOr(undefined)?.toHex() ?? null,
+    subject: credential.subject.toUtf8() as AssetDidUri,
+  }
```

Comment:
I think destructuring the properties from credential will make this code more readable

Reply:
Like this? https://github.com/KILTprotocol/sdk-js/pull/677/commits/023ad1c369410f4be90a93e0c4363bf2b8759885#diff-3228260bd7a8a1c0fb1e0b94efbc0b692518737b171dda86329b57db740703bbR81

Reply:
Yes, and `subject.toUtf8()` is done twice

Reply:
Good point: https://github.com/KILTprotocol/sdk-js/pull/677/commits/31a33366db67f7deb50aecba5e2c9f61c0df3c5e